### DISCUSSION TITLE
SQLAlchemy connection drop-outs on Flask + Gunicorn

### DISCUSSION THREAD

**Discussion by dodancs** [sqlalchemy/sqlalchemy#7985] [Usage Questions]
### Describe the bug

This is an issue I have been facing for a very long time with SQLAlchemy in combination with `Flask` + `Gunicorn` + `gevent worker`/`tornado worker` and `psycopg2` or `pg8000` driver with PostgreSQL 13 server.

I cannot really figure out when or why it happens and what is the cause of this. Basically what happens is that after any amount of time, when you do basic requests to your Flask API which request data from a PG database, the requests stop going through with messages like:

- `psycopg2.DatabaseError: error with status PGRES_TUPLES_OK and no message from the libpq`
- `sqlalchemy.exc.ResourceClosedError: This result object does not return rows. It has been closed automatically.`
- `lost synchronization with server: got message type "t", length 65542`
- `WARNING: there is already a transaction in progress`
- `DatabaseError: (psycopg2.DatabaseError) error with no message from the libpq`
- or just an uncaught exception `ERROR:tornado.application:Uncaught exception` (stacktrace in the error section)
- and others

I run everything in docker, with the following libraries and stuff:
```Dockerfile
FROM python:3.10-alpine

WORKDIR /app

RUN mkdir -p /app/storage/config

# install required packages
RUN apk add --no-cache postgresql-libs libpq curl tzdata libjpeg libpng libexif firefox ttf-freefont openssl ca-certificates

# install python requirements
COPY ./requirements.txt /app/requirements.txt
RUN \
    apk add --no-cache --virtual .build-deps gcc g++ make musl-dev python3-dev libffi-dev postgresql-dev openssl-dev && \
    pip install --no-cache-dir -r /app/requirements.txt && \
    rm -f /app/requirements.txt && \
    apk --purge del .build-deps

# setup environment variables
ENV PYTHONPATH=/app
ENV MODULE_NAME main
ENV VARIABLE_NAME app
ENV GUNICORN_CMD_ARGS --timeout 30 --threads 4
ENV WORKERS_PER_CORE 4
```

My `requirements.txt`:

```
alembic==1.7.7
blinker==1.4
dnstwist==20220131
Flask==2.1.1
Flask-Cors==3.0.10
Flask-JWT-Extended==4.3.1
Flask-Limiter==2.3.2
Flask-Migrate==3.1.0
Flask-RESTful==0.3.9
Flask-Script==2.0.6
Flask-SQLAlchemy==2.5.1
gunicorn==20.1.0
kafka-python==2.0.2
marshmallow==3.15.0
marshmallow-enum==1.5.1
pg8000==1.26.1
pillow==9.1.0
redis==4.1.4
requests==2.27.1
schedule==1.1.0
selenium==4.1.3
tornado==6.1
```

The `postgresql.conf` looks like so:

```ini
max_connections = 128
shared_buffers = 2GB
effective_cache_size = 6GB
maintenance_work_mem = 512MB
checkpoint_completion_target = 0.9
wal_buffers = 16MB
default_statistics_target = 100
random_page_cost = 4
effective_io_concurrency = 2
work_mem = 4MB
min_wal_size = 1GB
max_wal_size = 4GB
max_worker_processes = 8
max_parallel_workers_per_gather = 4
max_parallel_workers = 8
max_parallel_maintenance_workers = 4
```

Flask app config:

```python
SQLALCHEMY_DATABASE_URI = f'postgresql+pg8000://{DB_USER}:{DB_PASSWORD}@{DB_URL}/{DB_DATABASE}'
SQLALCHEMY_TRACK_MODIFICATIONS = False
SQLALCHEMY_ENGINE_OPTIONS = {
    'pool_size': 32,
    'pool_recycle': 60,
    'pool_pre_ping': True,
    'pool_timeout': 30
}
```

DB manager:

```python
from flask_sqlalchemy import SQLAlchemy
import logging
import os

if os.getenv('SQLALCHEMY_LOGGING', 'false').lower() == 'true':
    logging.basicConfig()
    logger = logging.getLogger('sqlalchemy.engine')
    logger.setLevel(logging.INFO)

db = SQLAlchemy()


def initialize(app):
    db.init_app(app)
```

### To Reproduce
Have any model, with some cross-relationships, and just return all() for example.
```python


from managers.db_manager import db
return db.session.query(cls).order_by(cls.created).all()


```
Usually, the biggest queries trigger the chain of reactions which leads to broken requests and error 500 from the Flask API. In my Flask app, I create a DB session for each worker.

### Error

```
core_1           | [2022-05-03 16:10:17 +0200] [1] [CRITICAL] WORKER TIMEOUT (pid:19)
core_1           | ERROR:tornado.application:Uncaught exception
core_1           | Traceback (most recent call last):
core_1           |   File "/usr/local/lib/python3.10/site-packages/tornado/http1connection.py", line 273, in _read_message
core_1           |     delegate.finish()
core_1           |   File "/usr/local/lib/python3.10/site-packages/tornado/httpserver.py", line 280, in finish
core_1           |     self.request_callback(self.request)
core_1           |   File "/usr/local/lib/python3.10/site-packages/tornado/wsgi.py", line 113, in __call__
core_1           |     app_response = self.wsgi_application(
core_1           |   File "/usr/local/lib/python3.10/site-packages/flask/app.py", line 2095, in __call__
core_1           |     return self.wsgi_app(environ, start_response)
core_1           |   File "/usr/local/lib/python3.10/site-packages/flask/app.py", line 2077, in wsgi_app
core_1           |     response = self.full_dispatch_request()
core_1           |   File "/usr/local/lib/python3.10/site-packages/flask/app.py", line 1523, in full_dispatch_request
core_1           |     rv = self.dispatch_request()
core_1           |   File "/usr/local/lib/python3.10/site-packages/flask/app.py", line 1509, in dispatch_request
core_1           |     return self.ensure_sync(self.view_functions[rule.endpoint])(**req.view_args)
core_1           |   File "/usr/local/lib/python3.10/site-packages/flask_restful/__init__.py", line 467, in wrapper
core_1           |     resp = resource(*args, **kwargs)
core_1           |   File "/usr/local/lib/python3.10/site-packages/flask/views.py", line 84, in view
core_1           |     return current_app.ensure_sync(self.dispatch_request)(*args, **kwargs)
core_1           |   File "/usr/local/lib/python3.10/site-packages/flask_restful/__init__.py", line 582, in dispatch_request
core_1           |     resp = meth(*args, **kwargs)
core_1           |   File "/app/managers/auth_manager.py", line 54, in wrapper
core_1           |     if TokenBlacklist.is_invalid(jwt_value):
core_1           |   File "/app/model/token_blacklist.py", line 23, in is_invalid
core_1           |     return db.session.query(cls.token).filter_by(token=token).first()
core_1           |   File "/usr/local/lib/python3.10/site-packages/sqlalchemy/orm/query.py", line 2819, in first
core_1           |     return self.limit(1)._iter().first()
core_1           |   File "/usr/local/lib/python3.10/site-packages/sqlalchemy/orm/query.py", line 2903, in _iter
core_1           |     result = self.session.execute(
core_1           |   File "/usr/local/lib/python3.10/site-packages/sqlalchemy/orm/session.py", line 1696, in execute
core_1           |     result = conn._execute_20(statement, params or {}, execution_options)
core_1           |   File "/usr/local/lib/python3.10/site-packages/sqlalchemy/engine/base.py", line 1631, in _execute_20
core_1           |     return meth(self, args_10style, kwargs_10style, execution_options)
core_1           |   File "/usr/local/lib/python3.10/site-packages/sqlalchemy/sql/elements.py", line 325, in _execute_on_connection
core_1           |     return connection._execute_clauseelement(
core_1           |   File "/usr/local/lib/python3.10/site-packages/sqlalchemy/engine/base.py", line 1490, in _execute_clauseelement
core_1           |     compiled_sql, extracted_params, cache_hit = elem._compile_w_cache(
core_1           |   File "/usr/local/lib/python3.10/site-packages/sqlalchemy/sql/elements.py", line 523, in _compile_w_cache
core_1           |     compiled_sql = self._compiler(
core_1           |   File "/usr/local/lib/python3.10/site-packages/sqlalchemy/sql/elements.py", line 558, in _compiler
core_1           |     return dialect.statement_compiler(dialect, self, **kw)
core_1           |   File "/usr/local/lib/python3.10/site-packages/sqlalchemy/sql/compiler.py", line 776, in __init__
core_1           |     Compiled.__init__(self, dialect, statement, **kwargs)
core_1           |   File "/usr/local/lib/python3.10/site-packages/sqlalchemy/sql/compiler.py", line 451, in __init__
core_1           |     self.string = self.process(self.statement, **compile_kwargs)
core_1           |   File "/usr/local/lib/python3.10/site-packages/sqlalchemy/sql/compiler.py", line 486, in process
core_1           |     return obj._compiler_dispatch(self, **kwargs)
core_1           |   File "/usr/local/lib/python3.10/site-packages/sqlalchemy/sql/visitors.py", line 82, in _compiler_dispatch
core_1           |     return meth(self, **kw)
core_1           |   File "/usr/local/lib/python3.10/site-packages/sqlalchemy/sql/compiler.py", line 3265, in visit_select
core_1           |     compile_state = select_stmt._compile_state_factory(
core_1           |   File "/usr/local/lib/python3.10/site-packages/sqlalchemy/sql/base.py", line 510, in create_for_statement
core_1           |     return klass.create_for_statement(statement, compiler, **kw)
core_1           |   File "/usr/local/lib/python3.10/site-packages/sqlalchemy/orm/context.py", line 696, in create_for_statement
core_1           |     _QueryEntity.to_compile_state(
core_1           |   File "/usr/local/lib/python3.10/site-packages/sqlalchemy/orm/context.py", line 2463, in to_compile_state
core_1           |     _ColumnEntity._for_columns(
core_1           |   File "/usr/local/lib/python3.10/site-packages/sqlalchemy/orm/context.py", line 2809, in _for_columns
core_1           |     _ORMColumnEntity(
core_1           |   File "/usr/local/lib/python3.10/site-packages/sqlalchemy/orm/context.py", line 3007, in __init__
core_1           |     _entity._post_inspect
core_1           |   File "/usr/local/lib/python3.10/site-packages/sqlalchemy/util/langhelpers.py", line 1180, in __get__
core_1           |     obj.__dict__[self.__name__] = result = self.fget(obj)
core_1           |   File "/usr/local/lib/python3.10/site-packages/sqlalchemy/orm/mapper.py", line 2119, in _post_inspect
core_1           |     self._check_configure()
core_1           |   File "/usr/local/lib/python3.10/site-packages/sqlalchemy/orm/mapper.py", line 1896, in _check_configure
core_1           |     _configure_registries({self.registry}, cascade=True)
core_1           |   File "/usr/local/lib/python3.10/site-packages/sqlalchemy/orm/mapper.py", line 3400, in _configure_registries
core_1           |     with _CONFIGURE_MUTEX:
core_1           |   File "/usr/local/lib/python3.10/site-packages/gunicorn/workers/base.py", line 203, in handle_abort
core_1           |     sys.exit(1)
core_1           | SystemExit: 1
core_1           | [2022-05-03 16:10:18 +0200] [19] [INFO] Worker exiting (pid: 19)
core_1           | [2022-05-03 16:10:18 +0200] [1] [WARNING] Worker with pid 19 was terminated due to signal 9
core_1           | [2022-05-03 16:10:18 +0200] [2357] [INFO] Booting worker with pid: 2357
``` 

### Versions

- OS: Alpine 3.15
- Python: 3.10
- SQLAlchemy: 1.4.36
- Flask-SQLAlchemy: 2.5.1
- Database: Postgresql [postgres:13-alpine](https://github.com/docker-library/postgres/blob/e483778176ca34bcbe83ee17000820d4f6e64c28/13/alpine/Dockerfile)
- DBAPI (eg: psycopg, cx_oracle, mysqlclient):  `psycopg2-binary==2.9.3`  //  `pg8000==1.26.1` (happens on both)
- Flask: 2.1.1
- Gunicorn: 20.1.0

### Also versions:

- OS: Alpine 3.8
- Python: 3.7
- SQLAlchemy: 1.3.11
- Flask-SQLAlchemy: 2.4.1
- Database: Postgresql [postgres:13-alpine](https://github.com/docker-library/postgres/blob/e483778176ca34bcbe83ee17000820d4f6e64c28/13/alpine/Dockerfile)
- DBAPI (eg: psycopg, cx_oracle, mysqlclient): `psycopg2-binary==2.8.4`
- Flask: 1.1.1
- Gunicorn: 20.0.4

### Additional context

I can share the whole of my code after the 16th of May if the supplied information is not enough.

**Comment by zzzeek** [comment]
hey there -

this doesnt illustrate any bug in SQLAlchemy, you would need to debug the concurrency behavior and network configuration of your application.   In particular when gunicorn is converting applications to "implicit async" this is a frequent cause of connectivity issues that involve the interaction between the database driver and the method by which gunicorn achieves this.   you would get more specific help from gunicorn mailing lists or groups on debugging this sort of issue.

**Reply by dodancs** [reply]
Hello, thanks for the fast reply.

I am really confused as I tried many different remedies to this, and also tried to only create one DB session for the whole application. This resulted in still the same issues, and I thought that as [JDiMatteo suggested](https://stackoverflow.com/a/43648900/1709032) this might be related to the Flask app improperly sharing the connection. Then I found [this article](https://davidcaron.dev/sqlalchemy-multiple-threads-and-processes/#:~:text=Every%20pool%20implementation%20in%20SQLAlchemy,will%20also%20be%20thread%2Dsafe.) stating a Queue Pool should manage the connections nicely in a thread-safe manner, and it is also used in a Flask app running on gunicorn. So I assumed that the issue probably is not with the way Flask accesses the DB session.

I will try to get into contact with the gunicorn team and figure what is going on, as I have observed these issues in multiple projects I work on or have contributed to.

Thank you for now!

**Reply by zzzeek** [reply]
> Hello, thanks for the fast reply.
> 
> I am really confused as I tried many different remedies to this, and also tried to only create one DB session for the whole application. 

that would not be a good idea.   a Session is a local, not-concurrent-safe object that corresponds to a single database transaction in progress.  so create Session objects per web request.  See the guidelines at https://docs.sqlalchemy.org/en/14/orm/session_basics.html#is-the-session-thread-safe 

> This resulted in still the same issues, and I thought that as [JDiMatteo suggested](https://stackoverflow.com/a/43648900/1709032) this might be related to the Flask app improperly sharing the connection. Then I found [this article](https://davidcaron.dev/sqlalchemy-multiple-threads-and-processes/#:~:text=Every%20pool%20implementation%20in%20SQLAlchemy,will%20also%20be%20thread%2Dsafe.) stating a Queue Pool should manage the connections nicely in a thread-safe manner, 

the QueuePool is threadsafe to the degree that it is a registry for idle connections.  however, when you check out a DBAPI connection from the pool, the connection itself is not threadsafe.

> and it is also used in a Flask app running on gunicorn. So I assumed that the issue probably is not with the way Flask accesses the DB session.

it's likely not flask by itself but some pattern you have configured within your application.  the Session itself is not threadsafe but note that also all the ORM objects that are associated with that Session are also part of that same transaction.  make sure nothing from the ORM Session or any instances of ORM objects associated with a Session are shared to other requests or to any kind of globally shared state at any time.

> 
> I will try to get into contact with the gunicorn team and figure what is going on, as I have observed these issues in multiple projects I work on or have contributed to.
> 
> Thank you for now!

good luck!

**Reply by dodancs** [reply]
> the QueuePool is threadsafe to the degree that it is a registry for idle connections. however, when you check out a DBAPI connection from the pool, the connection itself is not threadsafe.

To "check out a DPAPI connection from the pool" means start a transaction with `db.session.begin()` or access an object from the database with a query?

> it's likely not flask by itself but some pattern you have configured within your application. the Session itself is not threadsafe but note that also all the ORM objects that are associated with that Session are also part of that same transaction. make sure nothing from the ORM Session or any instances of ORM objects associated with a Session are shared to other requests or to any kind of globally shared state at any time.

This makes sense. I think I have Gunicorn configured to create 8 workers and each worker has 4 threads. So the context and therefore DB session is shared across all of the threads of a particular worker. This means then that if a user executes an API call on multiple of the worker threads at the same time, issues can arise as in this context, the code would try to access the same DB session, which, as you mentioned, is not good.

I'll try to look deeper into it and will have to probably ditch all threads and only use multiple workers. The downside is that that would mean only a small number of clients would be able to connect at the same time, since one worker could process only one HTTP request. Or use the second approach to always create new connections to the database for each request as you suggested.

**Reply by zzzeek** [reply]
> > the QueuePool is threadsafe to the degree that it is a registry for idle connections. however, when you check out a DBAPI connection from the pool, the connection itself is not threadsafe.
> 
> To "check out a DPAPI connection from the pool" means start a transaction with `db.session.begin()` or access an object from the database with a query?

We're first assuming you arent using the legacy "autocommit=True" Session flag.    Assuming that, the Session will check out the connection from the pool the first time it is asked to emit SQL.   This happens whether or not you said "session.begin()".    Session.begin() itself does not actually check out the connection since DBAPI works based on autobegin itself, there's nothing to emit to "begin" a transaction.


> 
> I'll try to look deeper into it and will have to probably ditch all threads and only use multiple workers. The downside is that that would mean only a small number of clients would be able to connect at the same time, since one worker could process only one HTTP request. Or use the second approach to always create new connections to the database for each request as you suggested.

that shouldn't be necessary, basically the main thing is that you make a new Session when there's a new web request, use it only in that request, then at the end of the request, close it.  that's really all you have to do.

**Reply by CaselIT** [reply]
As mike suggested, the preferred approach in web-servers is to create a session for each web request. I think flask-sqlalchemy is configured by default to do that for you.

I noticed that the application is  also using tornado, and that the strack trace is for tornato. This seems strange to use both flask and tornado, since they use a different concurrency model, so I'm gessing that it's a contributing factor on this problem.

What are you using tornado for?

Can't say I have experience on this, but you should probably not share the same session between tornato and flask

**Reply by dodancs** [reply]
@CaselIT I use tornado as the worker class for gunicorn. It is suggested on different websites (and also Flask itself says you should not use its built-in WSGI server) that for a production deployment you should use (preferably) gunicorn and some type of worker (`gevent`, `tornado`, `eventlet`, etc.). So I was using `gevent`, but thought the issue was with it so I switched to `tornado`. Ref: https://docs.gunicorn.org/en/stable/settings.html#worker-class

So basically Tornado is the web server used instead of the built-in WSGI server that Flask has.

**Reply by CaselIT** [reply]
You should have probably mentioned it in the opening issue.

When using tornato or gevent or other kind of "green" thread linrary, you should not share state among the different webrequest, since even "thread" safe code is not safe when using with green threads.

You could try to use the default sync worker of gunicorn so see if that fixes the issue.
If it does then you need to ensure that the session are not shared among different requests.

You can also use the default sync worker for gunicorn, it's perflectly fine unless you have a really large amount of concurrent clients

**Reply by CaselIT** [reply]
Also please note that the gunicorn docs advise against using tornado to serve wsgi application (like flask) https://docs.gunicorn.org/en/stable/design.html#tornado-workers

As mentioned above I have no direct experience here, so it may work well for you in the end anyway

**Reply by dodancs** [reply]
The issue why I could not use the default `sync` worker in one of the projects is the requirement for long-lasting HTTP sessions with keepalive (for [SSE](https://en.wikipedia.org/wiki/Server-sent_events#:~:text=Server%2DSent%20Events%20(SSE),client%20connection%20has%20been%20established.)).

I also just noticed some information about workers on the gunicorn wiki, stating that for greenlet and other "green" workers, I should use a special fork of the psycopg library - https://github.com/psycopg/psycogreen/. Maybe this also contributes to the usability/stability of the multi-threaded application.

**Reply by CaselIT** [reply]
If you specify threads the default is gthread instead of sync worker, that should support keepalive according to the docs
- https://docs.gunicorn.org/en/stable/design.html#asyncio-workers
- https://docs.gunicorn.org/en/stable/settings.html#threads

**Reply by olliemath** [reply]
As a middle-ground, the gevent worker works quite well with SQLAlchemy + Flask. 

For web dev, gevent feels very much like working with the regular threaded worker, because it does a lot of work in patching things (e.g. thread-local storage) to act like real threads.

If using `psycopg2` you also need the `psycogreen` library or it will block on database requests (although this may not be a problem if these are expected to be quick).

**Reply by seanabraham** [reply]
@dodancs I know this is years later, but did you ever get to the bottom of what specifically was causing the issue you were seeing? I'm seeing something similar (after many years of relative stability in a similar configuration) and wondering what might have happened and what ultimately resolved the issue for you.

**Reply by siddsengupta** [reply]
@dodancs - I am facing the same issue and I was wondering if you were able to resolve this. Thanks.
